tools/java/org/unicode/cldr/util/XMLSource.java

```diff
@@ -1747,4 +1764,711 @@ public DtdType getDtdType() {
         }
         return null;
     }
+
+    private DtdType dtdType;
+    public DtdType getSimpleXMLSourceDtdType() {
+        return dtdType;
+    }
+
+    public void setDtdType(DtdType dtdType) {
+        this.dtdType = dtdType;
+    }
+
+    private static class XMLSourceCacheKey {
+        String localeId;
+        List<File> dirs;
+        DraftStatus minimalDraftStatus;
+        int hashCode;
+        public XMLSourceCacheKey(String localeId, List<File> dirs, DraftStatus minimalDraftStatus) {
+            this.localeId = localeId;
+            this.dirs = dirs;
+            this.minimalDraftStatus = minimalDraftStatus;
+            this.hashCode = Objects.hash(this.localeId, this.dirs, this.minimalDraftStatus);
+        }
+
+        @Override
+        public int hashCode() {
+            return hashCode;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            XMLSourceCacheKey other = (XMLSourceCacheKey) obj;
+            if (!Objects.equals(dirs, other.dirs)) {
+                return false;
+            }
+            if (minimalDraftStatus != other.minimalDraftStatus) {
+                return false;
+            }
+            if (localeId == null) {
+                if (other.localeId != null) {
+                    return false;
+                }
+            } else if (!localeId.equals(other.localeId)) {
+                return false;
+            }
+            return true;
+        }
+    }
+
+    private static final int CACHE_LIMIT = 500;
+    private static LoadingCache<XMLSourceCacheKey, XMLSource> cache = CacheBuilder.newBuilder()
+        .maximumSize(CACHE_LIMIT)
+        .softValues()
+        .build(
+            new CacheLoader<XMLSourceCacheKey, XMLSource>() {
+                @Override
+                public XMLSource load(XMLSourceCacheKey key) {
+                    return makeXMLSource(key);
+                }
+            });
+
+    public static XMLSource makeXMLSource(XMLSourceCacheKey key) {
+        XMLSource source = null;
+        List<XMLSource> list = new ArrayList<>();
+        if (key.dirs.size() == 1) {
+            File file = new File(key.dirs.get(0), key.localeId + ".xml");
+            source = loadXMLFile(file, key.localeId, key.minimalDraftStatus);
+            source.freeze();
+            return source;
+        }
+
+        List<File> dirList = new ArrayList<>();
+        for (File dir: key.dirs) {
+            dirList.clear();
+            dirList.add(dir);
+            XMLSourceCacheKey singleKey = new XMLSourceCacheKey(key.localeId, dirList, key.minimalDraftStatus);
+            XMLSource singleSource = null;
+            singleSource = cache.getUnchecked(singleKey);
+            list.add(singleSource);
```

Comment:
If `singleKey` is no in `cache` then we add `null` to `list`?

Reply:
will check null before put into cache